source/common/thread_local/thread_local_impl.cc

```diff
@@ -93,24 +112,24 @@ void InstanceImpl::setThreadLocal(uint32_t index, ThreadLocalObjectSharedPtr obj
 
 void InstanceImpl::shutdownGlobalThreading() {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
   shutdown_ = true;
 }
 
 void InstanceImpl::shutdownThread() {
+  ASSERT(shutdown_);
+
   // Destruction of slots is done in *reverse* order. This is so that filters and higher layer
   // things that are built on top of the cluster manager, stats, etc. will be destroyed before
   // more base layer things. It's possible this might need to become more complicated later but
-  // it's OK for now.
+  // it's OK for now. Note that this is always safe to do because:
```

Comment:
Yeah, makes sense. I like the idea of an allocation order list, can you add a TODO and some explanation? What I would really like to see is the TLS dependency graph that exists today, since this requires some global knowledge of the code that I don't have right now.